### PR TITLE
Support gbk Encoding To Fix #532 

### LINKED ISSUE #532: error 1366 on invalid character
I believe i have run into the same issue #290 ,But my table is gbk-encoded and contains utf8 characters.

For reproduce it , I add postpone file ,and update a row after row-copy complete .

Actually  the column of name's value is `涪陵北`. 

v1.0.40

**table structure**
```
CREATE TABLE `recur_gbk` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `Name` varchar(512) DEFAULT NULL,
  `addcol` varchar(255) DEFAULT NULL COMMENT '添加普通列测试',
  PRIMARY KEY (`id`)
) ENGINE=InnoDB AUTO_INCREMENT=32 DEFAULT CHARSET=gbk
```

**gh-ost command**
```
gh-ost --host=10.28.77.43 --database="ghostdb" --table="recur_gbk" --debug --alter="ADD COLUMN addcol1 VARCHAR(255) DEFAULT NULL COMMENT '添加普通列测试'" --concurrent-rowcount --conf="/usr/local/gh-ost.cnf" --port=55944 --execute --postpone-cut-over-flag-file="/tmp/r.postpone" --initially-drop-old-table
```

**log**
```
2018-01-02 19:03:44 INFO starting gh-ost 1.0.40
2018-01-02 19:03:44 INFO Migrating `ghostdb`.`recur_gbk`
2018-01-02 19:03:44 INFO connection validated on 10.28.77.43:55944
2018-01-02 19:03:44 INFO User has ALL privileges
2018-01-02 19:03:44 INFO binary logs validated on 10.28.77.43:55944
2018-01-02 19:03:44 INFO Restarting replication on 10.28.77.43:55944 to make sure binlog settings apply to replication thread
2018-01-02 19:03:44 DEBUG Replication restarted
2018-01-02 19:03:44 INFO Inspector initiated on SVR14883DE630:55944, version 5.7.18-ctrip-log
2018-01-02 19:03:44 INFO Table found. Engine=InnoDB
2018-01-02 19:03:44 DEBUG Estimated number of rows via STATUS: 19
2018-01-02 19:03:44 DEBUG Validated no foreign keys exist on table
2018-01-02 19:03:44 DEBUG Validated no triggers exist on table
2018-01-02 19:03:44 INFO Estimated number of rows via EXPLAIN: 19
2018-01-02 19:03:44 DEBUG Potential unique keys in recur_gbk: [PRIMARY (auto_increment): [id]; has nullable: false]
2018-01-02 19:03:44 INFO Recursively searching for replication master
2018-01-02 19:03:44 DEBUG Looking for master on 10.28.77.43:55944
2018-01-02 19:03:44 DEBUG Master of 10.28.77.43:55944 is 10.28.77.42:55944
2018-01-02 19:03:44 DEBUG Looking for master on 10.28.77.42:55944
2018-01-02 19:03:44 INFO Master found to be 10.28.77.42:55944
2018-01-02 19:03:44 INFO log_slave_updates validated on 10.28.77.43:55944
2018-01-02 19:03:44 INFO connection validated on 10.28.77.43:55944
2018-01-02 19:03:44 DEBUG Streamer binlog coordinates: SVR14883DE630-bin.001903:55888
2018/01/02 19:03:44 binlogsyncer.go:76: [info] create BinlogSyncer with config 10.28.77.43:55944[replica-server-id:99999]
2018-01-02 19:03:44 INFO Connecting binlog streamer at SVR14883DE630-bin.001903:55888
2018/01/02 19:03:44 binlogsyncer.go:242: [info] begin to sync binlog from position (SVR14883DE630-bin.001903, 55888)
2018/01/02 19:03:44 binlogsyncer.go:135: [info] register slave for master server 10.28.77.43:55944
2018/01/02 19:03:44 binlogsyncer.go:569: [info] rotate to (SVR14883DE630-bin.001903, 55888)
2018-01-02 19:03:44 DEBUG Beginning streaming
2018-01-02 19:03:44 INFO rotate to next log name: SVR14883DE630-bin.001903
2018-01-02 19:03:44 INFO connection validated on 10.28.77.42:55944
2018-01-02 19:03:44 INFO connection validated on 10.28.77.42:55944
2018-01-02 19:03:44 INFO will use time_zone='SYSTEM' on applier
2018-01-02 19:03:44 INFO Examining table structure on applier
2018-01-02 19:03:44 INFO Applier initiated on SVR14306DE630:55944, version 5.7.18-ctrip-log
2018-01-02 19:03:44 INFO Droppping table `ghostdb`.`_recur_gbk_del`
2018-01-02 19:03:44 INFO Table dropped
2018-01-02 19:03:44 INFO Droppping table `ghostdb`.`_recur_gbk_ghc`
2018-01-02 19:03:44 INFO Table dropped
2018-01-02 19:03:44 INFO Creating changelog table `ghostdb`.`_recur_gbk_ghc`
2018-01-02 19:03:44 INFO Changelog table created
2018-01-02 19:03:44 INFO Creating ghost table `ghostdb`.`_recur_gbk_gho`
2018-01-02 19:03:44 INFO Ghost table created
2018-01-02 19:03:44 INFO Altering ghost table `ghostdb`.`_recur_gbk_gho`
2018-01-02 19:03:44 DEBUG ALTER statement: alter /* gh-ost */ table `ghostdb`.`_recur_gbk_gho` ADD COLUMN addcol1 VARCHAR(255) DEFAULT NULL COMMENT '添加普通列测试'
2018-01-02 19:03:44 INFO Ghost table altered
2018-01-02 19:03:44 INFO Waiting for ghost table to be migrated. Current lag is 0s
2018-01-02 19:03:44 INFO Intercepted changelog state GhostTableMigrated
2018-01-02 19:03:44 INFO Handled changelog state GhostTableMigrated
2018-01-02 19:03:44 DEBUG ghost table migrated
2018-01-02 19:03:44 DEBUG Potential unique keys in _recur_gbk_gho: [PRIMARY (auto_increment): [id]; has nullable: false]
2018-01-02 19:03:44 INFO Chosen shared unique key is PRIMARY
2018-01-02 19:03:44 INFO Shared columns are id,Name,addcol
2018-01-02 19:03:44 INFO Listening on unix socket file: /tmp/gh-ost.ghostdb.recur_gbk.sock
2018-01-02 19:03:44 DEBUG Reading migration range according to key: PRIMARY
2018-01-02 19:03:44 INFO Migration min values: [1]
2018-01-02 19:03:44 DEBUG Reading migration range according to key: PRIMARY
2018-01-02 19:03:44 INFO Migration max values: [31]
2018-01-02 19:03:44 INFO Waiting for first throttle metrics to be collected
2018-01-02 19:03:44 INFO First throttle metrics collected
2018-01-02 19:03:44 DEBUG Operating until row copy is complete
Copy: 19/19 100.0%; Applied: 0; Backlog: 0/1000; Time: 1m0s(total), 1s(copy); streamer: SVR14883DE630-bin.001903:315958; State: postponing cut-over; ETA: due
2018-01-02 19:04:57 ERROR Error 1366: Incorrect string value: '\xB8\xA2\xC1\xEA\xB1\xB1' for column 'Name' at row 1; query=
 			update /* gh-ost `ghostdb`.`_recur_gbk_gho` */
 					`ghostdb`.`_recur_gbk_gho`
				set
					`id`=?, `Name`=?, `addcol`=?
				where
 					((`id` = ?))
 		; args=[1 ¸¢ª±± 1 1]
2018-01-02 19:04:58 ERROR Error 1366: Incorrect string value: '\xB8\xA2\xC1\xEA\xB1\xB1' for column 'Name' at row 1; query=
 			update /* gh-ost `ghostdb`.`_recur_gbk_gho` */
 					`ghostdb`.`_recur_gbk_gho`
				set
					`id`=?, `Name`=?, `addcol`=?
				where
 					((`id` = ?))
 		; args=[1 ¸¢ª±± 1 1]
```
> Thank you!

### PR DESCRIPTION
**Pull request by ceshihao**
To fix #532 

### Description

This PR adds `gbk` encoding support.

### CODE REVIEW

**Comment by ceshihao** [comment]
also related to #228

**Comment by shlomi-noach** [comment]
Thank you. Adding GBK will create a different problem as per https://github.com/go-sql-driver/mysql#interpolateparams. We would need to turn off `interpolateParams`, meaning we'd need to turn on prepared statements.

0. Is this fix only applicable when the migrated table has a column of `CHARACTER SET = GBK`?
0. Or is there a possibility that some `utf8` column would have `GBK` encoding?

If the former, then we can conditionally disable `interpolateParams` for migrations where `GBK` is present.

**Comment by ceshihao** [comment]
I add an option to indicate charset includes multibyte encoding which can be used to determine whether disable interpolateParams.

**Comment by shlomi-noach** [comment]
Do you think you have the answer though?

**Comment by ceshihao** [comment]
I think this PR only handles the case with `gbk` encoding columns.

For question 2, i am not sure about the solution.

**Comment by shlomi-noach** [comment]
Thank you. Please be advised that https://github.com/github/gh-ost/pull/526 is to be merged (if all goes well), and will affect your PR once merged, since it also touches (and moves) MySQL URI definitions.

**Comment by ceshihao** [comment]
> 2. Or is there a possibility that some utf8 column would have GBK encoding?

Do you have any idea about this case?

**Comment by shlomi-noach** [comment]
> Do you have any idea about this case? 

I don't...

File: go/mysql/connection.go [hunk 1]
```diff
@@ -49,12 +49,16 @@ func (this *ConnectionConfig) Equals(other *ConnectionConfig) bool {
...
 	hostname := this.Key.Hostname
 	var ip = net.ParseIP(hostname)
 	if (ip != nil) && (ip.To4() == nil) {
 		// Wrap IPv6 literals in square brackets
 		hostname = fmt.Sprintf("[%s]", hostname)
 	}
-	return fmt.Sprintf("%s:%s@tcp(%s:%d)/%s?interpolateParams=true&autocommit=true&charset=utf8mb4,utf8,latin1", this.User, this.Password, hostname, this.Key.Port, databaseName)
+	var multibyteCharset string
+	if includeMultibyte {
+		multibyteCharset = ",gbk,gb2312,big5,cp932,sjis"
```
Comment (shlomi-noach) [inline go/mysql/connection.go]: is `multibyteCharset` a correct description? `utf8` is also a multibyte charset. Is there a better definition?

File: go/mysql/connection.go [hunk 2]
```diff
@@ -49,12 +49,16 @@ func (this *ConnectionConfig) Equals(other *ConnectionConfig) bool {
...
+	}
+	return fmt.Sprintf("%s:%s@tcp(%s:%d)/%s?interpolateParams=%t&autocommit=true&charset=utf8mb4,utf8,latin1%s", this.User, this.Password, hostname, this.Key.Port, databaseName, !includeMultibyte, multibyteCharset)
```
Comment (shlomi-noach) [inline go/mysql/connection.go]: I'd create a variable called `interpolateParams` that happens to be set to `!includeMultibyte`, and use `interpolateParams` in the `Sprintf` statement. The logic to [not] use `interpolateParams` may end up considering additional inputs to the charset.

File: go/mysql/connection.go (hunk 1 above)
Comment (ceshihao) [reply go/mysql/connection.go]: How about `riskCharset` and `includeRiskCharset`?

File: go/mysql/connection.go (hunk 2 above)
Comment (cenkore) [reply go/mysql/connection.go]: Thanks
We need manual close `interpolateParams` if column charset with multibyte encodings like BIG5, CP932, GB2312, GBK or SJIS ?
`riskCharset` , Is there any risk for this fix ?

**Comment by ceshihao** [comment]
resolve conflict after #526 merged

File: go/mysql/connection.go (hunk 2 above)
Comment (shlomi-noach) [reply go/mysql/connection.go]: There's no risk for disabling `interpolateParams`, it is a matter of performance.

File: go/mysql/connection.go (hunk 1 above)
Comment (shlomi-noach) [reply go/mysql/connection.go]: sounds good.

**Comment by cenkore** [comment]
Any progress about this issue ?

**Comment by shlomi-noach** [comment]
@cenkore apologies for the delay! I hope to work on this in the next few days.

**Comment by shlomi-noach** [comment]
Sent to testing.

**Comment by shlomi-noach** [comment]
Can you please design a test (similar to https://github.com/github/gh-ost/tree/master/localtests/utf8mb4) that would express the support for the new changes?

**Comment by shlomi-noach** [comment]
Ignore the above. I will use https://github.com/github/gh-ost/issues/532 as a test.

**Comment by shlomi-noach** [comment]
Here's a funny thing: I added [this test](https://github.com/github/gh-ost/pull/533/files#diff-474fc96f896b5b449269e9325fc5991b) and it passes with a "normal" run, without adding the `--include-risk-charset` flag.

**Comment by ceshihao** [comment]
> Here's a funny thing: I added this test and it passes with a "normal" run, without adding the --include-risk-charset flag.

https://github.com/go-sql-driver/mysql/blob/master/dsn.go#L73

It seems that the validation in driver is on `Collation` instead of `Charset`.

**Comment by shlomi-noach** [comment]
Superseded by https://github.com/github/gh-ost/pull/560

**Pull request merged**